crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1093 @@
+    let mut decomposers: Vec<DecomposerElement> = Vec::new();
+
+    #[inline]
+    fn is_supercontrolled(op: &NormalOperation) -> bool {
+        match op.operation.matrix(&op.params) {
+            None => false,
+            Some(unitary_matrix) => {
+                let kak = TwoQubitWeylDecomposition::new_inner(unitary_matrix.view(), None, None)
+                    .unwrap();
+                relative_eq!(kak.a(), PI4) && relative_eq!(kak.c(), 0.0)
+            }
+        }
+    }
+
+    #[inline]
+    fn is_controlled(op: &NormalOperation) -> bool {
+        match op.operation.matrix(&op.params) {
+            None => false,
+            Some(unitary_matrix) => {
+                let kak = TwoQubitWeylDecomposition::new_inner(unitary_matrix.view(), None, None)
+                    .unwrap();
+                relative_eq!(kak.b(), 0.0) && relative_eq!(kak.c(), 0.0)
+            }
+        }
+    }
+
+    // Iterate over 1q and 2q supercontrolled basis, append TwoQubitBasisDecomposers
+    let supercontrolled_basis: IndexMap<&str, NormalOperation> = available_2q_basis
+        .clone()
+        .into_iter()
+        .filter(|(_, v)| is_supercontrolled(v))
+        .collect();
+
+    for basis_1q in &available_1q_basis {
+        for basis_2q in supercontrolled_basis.keys() {
+            let mut basis_2q_fidelity: f64 = match available_2q_props.get(basis_2q) {
+                Some(&(_, Some(e))) => 1.0 - e,
+                _ => 1.0,
+            };
+            if let Some(approx_degree) = approximation_degree {
+                basis_2q_fidelity *= approx_degree;
+            }
+            let gate = &supercontrolled_basis[basis_2q];
+            let decomposer = TwoQubitBasisDecomposer::new_inner(
+                gate.operation.name().to_owned(),
+                gate.operation.matrix(&gate.params).unwrap().view(),
+                basis_2q_fidelity,
+                basis_1q,
+                None,
+            )?;
+
+            decomposers.push(DecomposerElement {
+                decomposer: DecomposerType::TwoQubitBasisDecomposer(Box::new(decomposer)),
+                gate: gate.clone(),
+            });
+        }
+    }
+
+    // If our 2q basis gates are a subset of cx, ecr, or cz then we know TwoQubitBasisDecomposer
+    // is an ideal decomposition and there is no need to bother calculating the XX embodiments
+    // or try the XX decomposer
+    let available_basis_set: HashSet<&str> = available_2q_basis.keys().copied().collect();
+
+    #[inline]
+    fn check_goodbye(basis_set: &HashSet<&str>) -> bool {
+        !basis_set.iter().any(|gate| !GOODBYE_SET.contains(gate))
+    }
+
+    if check_goodbye(&available_basis_set) {
+        return Ok(Some(decomposers));
+    }
+
+    // Let's now look for possible controlled decomposers (i.e. XXDecomposer)
+    let controlled_basis: IndexMap<&str, NormalOperation> = available_2q_basis
+        .into_iter()
+        .filter(|(_, v)| is_controlled(v))
+        .collect();
```

Comment:
Here you can also skip some clones:
```suggestion
    // Let's now look for possible controlled decomposers (i.e. XXDecomposer)
    let controlled_basis: IndexMap<&str, NormalOperation> = available_2q_basis
        .iter()  
        .filter(|(_, v)| is_supercontrolled(v))  
        .map(|(k, v)| (*k, v.clone()))
        .collect();
```